storage/asynchbase/src/main/java/net/opentsdb/storage/Tsdb1xScanners.java

```diff
@@ -14,57 +14,965 @@
+    } else {
+      skip_nsun_tagvs = ((Tsdb1xHBaseDataStore) node.factory()).dynamicBoolean(Tsdb1xHBaseDataStore.SKIP_NSUN_TAGV_KEY);
+    }
+    if (query.hasKey(Tsdb1xHBaseDataStore.PRE_AGG_KEY)) {
+      pre_aggregate = query.getBoolean(config, Tsdb1xHBaseDataStore.PRE_AGG_KEY);
+    } else {
+      pre_aggregate = false;
+    }
+    if (query.hasKey(Tsdb1xHBaseDataStore.FUZZY_FILTER_KEY)) {
+      enable_fuzzy_filter = query.getBoolean(config, Tsdb1xHBaseDataStore.FUZZY_FILTER_KEY);
+    } else {
+      enable_fuzzy_filter = ((Tsdb1xHBaseDataStore) node.factory()).dynamicBoolean(Tsdb1xHBaseDataStore.FUZZY_FILTER_KEY);
+    }
+    if (query.hasKey(Tsdb1xHBaseDataStore.REVERSE_KEY)) {
+      reverse_scan = query.getBoolean(config, Tsdb1xHBaseDataStore.REVERSE_KEY);
+    } else {
+      reverse_scan = ((Tsdb1xHBaseDataStore) node.factory()).dynamicBoolean(Tsdb1xHBaseDataStore.REVERSE_KEY);
+    }
+    if (query.hasKey(Tsdb1xHBaseDataStore.MAX_MG_CARDINALITY_KEY)) {
+      max_multi_get_cardinality = query.getInt(config, Tsdb1xHBaseDataStore.MAX_MG_CARDINALITY_KEY);
+    } else {
+      max_multi_get_cardinality = ((Tsdb1xHBaseDataStore) node.factory()).dynamicInt(Tsdb1xHBaseDataStore.MAX_MG_CARDINALITY_KEY);
+    }
+    
+    if (node.schema().rollupConfig() != null && rollup_usage != RollupUsage.ROLLUP_RAW) {
+      Downsampler ds = query.getMetrics().get(0).getDownsampler();
+      if (ds != null) {
+        rollup_intervals = node.schema().rollupConfig().getRollupInterval(
+            DateTime.parseDuration(ds.getInterval()) / 1000, ds.getInterval());
+      } else if (query.getTime().getDownsampler() != null) {
+        ds = query.getTime().getDownsampler();
+        rollup_intervals = node.schema().rollupConfig().getRollupInterval(
+            DateTime.parseDuration(ds.getInterval()) / 1000, ds.getInterval());
+      } else {
+        rollup_intervals = null;
+      }
+      
+      if (ds != null) {
+        rollup_aggregation = ds.getAggregator();
+      } else {
+        rollup_aggregation = null;
+      }
+    } else {
+      rollup_intervals = null;
+      rollup_aggregation = null;
+    }
+    
+    if (Strings.isNullOrEmpty(query.getMetrics().get(0).getAggregator()) ) {
+      rollup_group_by = query.getTime().getAggregator();
+    } else {
+      rollup_group_by = query.getMetrics().get(0).getAggregator();
+    }
+  }
+  
+  /**
+   * Call to fetch the next set of data from the scanners. 
+   * <b>WARNING:</b> Do not call this from the parent node without receiving
+   * a response first. Only one call can be outstanding at a time.
+   * 
+   * @param result A non-null result.
+   * @param span An optional span.
+   * @throws IllegalArgumentException if the result is null.
+   */
+  public void fetchNext(final Tsdb1xQueryResult result, final Span span) {
+    if (result == null) {
+      throw new IllegalArgumentException("Result must be initialized");
+    }
+    
+    synchronized (this) {
+      if (current_result != null) {
```

Comment:
Are you expecting to see the memory effects from other threads here? That is, could another thread set `current_result` to `null`?

Reply:
Another thread could and they need to be blocked. It would mean there was a bug upstream.